File: ios/Runner/ParakeetNative.swift
import Foundation
import Accelerate

class ParakeetNative {
    
    private var isModelLoaded = false
    private var modelPath: String?
    
    // ONNX Runtime session (would use onnxruntime-objc framework)
    // For this example, we'll show the structure
    
    private let SAMPLE_RATE = 16000
    private let N_MELS = 80
    
    func loadModel(modelPath: String) -> Bool {
        guard FileManager.default.fileExists(atPath: modelPath) else {
            print("Model file not found: \(modelPath)")
            return false
        }
        
        do {
            // Initialize ONNX Runtime session
            // let session = try ORTSession(path: modelPath)
            
            self.modelPath = modelPath
            self.isModelLoaded = true
            
            print("STT model loaded successfully")
            return true
        } catch {
            print("Failed to load STT model: \(error)")
            return false
        }
    }
    
    func transcribe(audioData: Data) throws -> String {
        guard isModelLoaded else {
            throw NSError(domain: "ParakeetNative", code: 1, userInfo: [NSLocalizedDescriptionKey: "Model not loaded"])
        }
        
        // Convert PCM data to float array
        let audioFloats = pcmDataToFloatArray(audioData)
        
        // Prepare input tensor
        // let inputTensor = try ORTTensor(...)
        
        // Run inference
        // let outputs = try session.run(withInputs: ["audio_signal": inputTensor])
        
        // Get token IDs from output
        // let tokenIds = outputs["tokens"] as? [Int64] ?? []
        
        // Decode tokens to text
        // let transcript = decodeTokens(tokenIds)
        
        // For now, return placeholder
        let transcript = "Transcription placeholder"
        
        return transcript
    }
    
    private func pcmDataToFloatArray(_ data: Data) -> [Float] {
        var floats = [Float](repeating: 0, count: data.count / 2)
        
        data.withUnsafeBytes { (bytes: UnsafeRawBufferPointer) in
            let int16Buffer = bytes.bindMemory(to: Int16.self)
            
            for i in 0..<int16Buffer.count {
                // Normalize to [-1.0, 1.0]
                floats[i] = Float(int16Buffer[i]) / 32768.0
            }
        }
        
        return floats
    }
    
    private func decodeTokens(_ tokenIds: [Int64]) -> String {
        // TODO: Implement proper token decoding with vocabulary
        // This requires loading the tokenizer vocabulary file
        
        let vocab = loadVocabulary()
        var words: [String] = []
        
        for tokenId in tokenIds {
            if tokenId == 0 { break } // EOS token
            
            if let word = vocab[Int(tokenId)], word != "<pad>" {
                words.append(word)
            }
        }
        
        return words.joined(separator: " ").trimmingCharacters(in: .whitespaces)
    }
    
    private func loadVocabulary() -> [Int: String] {
        // TODO: Load actual vocabulary from file
        return [:]
    }
    
    func unload() {
        isModelLoaded = false
        modelPath = nil
        print("STT model unloaded")
    }
    
    func isLoaded() -> Bool {
        return isModelLoaded
    }
}

/*
 * To use ONNX Runtime on iOS, add to Podfile:
 * 
 * pod 'onnxruntime-objc', '~> 1.17.0'
 * 
 * Then import in bridging header:
 * #import <onnxruntime/onnxruntime.h>
 */

File: android/app/src/main/kotlin/ParakeetNative.kt
package com.example.nova_voice_assistant

import android.content.Context
import ai.onnxruntime.*
import java.nio.FloatBuffer
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * Native wrapper for Parakeet STT ONNX model
 * Handles speech-to-text transcription using ONNX Runtime
 */
class ParakeetNative(private val context: Context) {
    private var ortEnvironment: OrtEnvironment? = null
    private var ortSession: OrtSession? = null
    private var isModelLoaded = false
    
    // Model configuration
    private val SAMPLE_RATE = 16000
    private val N_MELS = 80
    
    companion object {
        init {
            System.loadLibrary("onnxruntime")
        }
    }
    
    /**
     * Load the Parakeet ONNX model from assets
     */
    fun loadModel(modelPath: String): Boolean {
        try {
            // Initialize ONNX Runtime environment
            ortEnvironment = OrtEnvironment.getEnvironment()
            
            // Create session options
            val sessionOptions = OrtSession.SessionOptions()
            sessionOptions.setIntraOpNumThreads(4)
            sessionOptions.setOptimizationLevel(OrtSession.SessionOptions.OptLevel.ALL_OPT)
            
            // Load model from assets or file path
            val modelBytes = if (modelPath.startsWith("assets/")) {
                context.assets.open(modelPath.removePrefix("assets/")).readBytes()
            } else {
                java.io.File(modelPath).readBytes()
            }
            
            // Create ONNX session
            ortSession = ortEnvironment?.createSession(modelBytes, sessionOptions)
            
            isModelLoaded = true
            return true
        } catch (e: Exception) {
            e.printStackTrace()
            return false
        }
    }
    
    /**
     * Transcribe audio bytes to text
     * Input: PCM 16-bit audio at 16kHz
     */
    fun transcribe(audioData: ByteArray): String {
        if (!isModelLoaded || ortSession == null) {
            throw IllegalStateException("Model not loaded")
        }
        
        try {
            // Convert PCM bytes to float array (normalize to [-1, 1])
            val audioFloats = pcmBytesToFloatArray(audioData)
            
            // Prepare input tensor
            val inputShape = longArrayOf(1, audioFloats.size.toLong())
            val inputTensor = OnnxTensor.createTensor(
                ortEnvironment,
                FloatBuffer.wrap(audioFloats),
                inputShape
            )
            
            // Run inference
            val inputs = mapOf("audio_signal" to inputTensor)
            val outputs = ortSession!!.run(inputs)
            
            // Get output tensor (token IDs)
            val outputTensor = outputs[0].value as Array<LongArray>
            val tokenIds = outputTensor[0]
            
            // Decode tokens to text (you'll need to implement a tokenizer)
            val transcript = decodeTokens(tokenIds)
            
            // Clean up
            inputTensor.close()
            outputs.close()
            
            return transcript
        } catch (e: Exception) {
            e.printStackTrace()
            throw RuntimeException("Transcription failed: ${e.message}")
        }
    }
    
    /**
     * Convert PCM 16-bit audio bytes to normalized float array
     */
    private fun pcmBytesToFloatArray(pcmBytes: ByteArray): FloatArray {
        val shortBuffer = ByteBuffer.wrap(pcmBytes)
            .order(ByteOrder.LITTLE_ENDIAN)
            .asShortBuffer()
        
        val floatArray = FloatArray(shortBuffer.remaining())
        for (i in floatArray.indices) {
            // Normalize to [-1.0, 1.0]
            floatArray[i] = shortBuffer.get(i) / 32768.0f
        }
        return floatArray
    }
    
    /**
     * Decode token IDs to text
     * This is a simplified version - you'll need to implement proper tokenizer
     */
    private fun decodeTokens(tokenIds: LongArray): String {
        // TODO: Implement proper tokenizer/vocabulary lookup
        // For now, this is a placeholder
        // You need to:
        // 1. Load vocabulary from tokenizer config
        // 2. Map token IDs to text
        // 3. Handle special tokens (pad, eos, etc.)
        
        val vocab = loadVocabulary()
        val words = mutableListOf<String>()
        
        for (tokenId in tokenIds) {
            if (tokenId == 0L) break // EOS token
            val word = vocab.getOrDefault(tokenId.toInt(), "<unk>")
            if (word != "<pad>") {
                words.add(word)
            }
        }
        
        return words.joinToString(" ").trim()
    }
    
    /**
     * Load vocabulary from assets
     * This needs to match the tokenizer used during model training
     */
    private fun loadVocabulary(): Map<Int, String> {
        // TODO: Load actual vocabulary file (vocab.json or similar)
        // This is a placeholder implementation
        return emptyMap()
    }
    
    /**
     * Unload model and free resources
     */
    fun unload() {
        ortSession?.close()
        ortSession = null
        isModelLoaded = false
    }
    
    /**
     * Check if model is loaded
     */
    fun isLoaded(): Boolean = isModelLoaded
}